### PR TITLE
implement scale messenger

### PR DESCRIPTION
**Pull request by fehiepsi**
This messenger is used to modify log density of sub-posteriors (rather reimplementing `potential_energy`, `initialize_model` for them),... One thing I concerned about is whether the logic at transform log_det is right, so I need your opinions about it.

### CODE REVIEW

File: numpyro/hmc_util.py [hunk 1]
```diff
@@ -696,16 +696,22 @@ def log_density(model, model_args, model_kwargs, params):
...
 def potential_energy(model, model_args, model_kwargs, inv_transforms):
     def _potential_energy(params):
         params_constrained = constrain_fn(inv_transforms, params)
-        log_joint = jax.partial(log_density, model, model_args, model_kwargs)(params_constrained)[0]
+        log_joint, model_trace = log_density(model, model_args, model_kwargs, params_constrained)
         for name, t in inv_transforms.items():
-            log_joint = log_joint + np.sum(t.log_abs_det_jacobian(params[name], params_constrained[name]))
+            t_log_det = np.sum(t.log_abs_det_jacobian(params[name], params_constrained[name]))
+            if 'scale' in model_trace[name]:
+                t_log_det = model_trace[name]['scale'] * t_log_det
```
Comment (fehiepsi) [inline numpyro/hmc_util.py]: do we need to scale `log_abs_det_jacobian`?

Comment (martinjankowiak) [reply numpyro/hmc_util.py]: this looks right. although i guess you're assuming scale is a float?

Comment (fehiepsi) [reply numpyro/hmc_util.py]: Yes. Do you know other usage cases, so I can address it or add a note for future reconsideration?

Comment (martinjankowiak) [reply numpyro/hmc_util.py]: this is fine. i think in the main repo we allow tensorial scaling through mask. you should probably explicitly say that scale is positive real-valued in the docstring and check that it's positive in the code

**Comment by neerajprad** [comment]
Feel free to merge pending any further comments from @martinjankowiak.
